pkg/controller/attach_detach.go

```diff
@@ -587,154 +572,160 @@ func reportError(azVolumeAttachment *azdiskv1beta2.AzVolumeAttachment, state azd
+	vInfo, ok := r.volumeInfos[volumeName]
+	if ok {
+		if synced, hasAzVolumeAttachment := vInfo.syncedAzVolumeAttachments[azVolumeAttachmentName]; hasAzVolumeAttachment && synced {
+			hasSynced = true
+			return volumeHandle, volumeContext, vInfo, hasSynced, nil
+		}
+		volumeHandle = vInfo.handle
+		volumeContext = vInfo.attributes
+	} else {
+		w.Logger().Infof("getting volumeHandle and volumeContext from PersistentVolume for volume (%s)", volumeName)
+		volumeHandle, volumeContext, err = r.getVolumeInfoFromPV(ctx, volumeName)
+		if err != nil {
+			w.Logger().Errorf(err, "failed to find volume info from PersistentVolume for volume (%s)", volumeName)
+			return "", nil, vInfo, hasSynced, err
+		}
+		vInfo = newVolumeInfo(volumeHandle, volumeContext)
+		r.volumeInfos[volumeName] = vInfo
+	}
+	return volumeHandle, volumeContext, vInfo, hasSynced, nil
+}
 
-				azVolumeAttachment, err = r.azClient.DiskV1beta2().AzVolumeAttachments(r.config.ObjectNamespace).Update(ctx, azVolumeAttachment, metav1.UpdateOptions{})
-				if err != nil {
-					w.Logger().Errorf(err, "failed to update AzVolumeAttachment (%s) for volume (%s) and node (%s): %v", azVolumeAttachmentName, *pvName, nodeName, err)
-					return syncedVolumeAttachments, volumesToSync, err
-				}
-			} else {
-				statusUpdateRequired = false
-			}
+func newVolumeInfo(volumeHandle string, volumeContext map[string]string) *volumeInfo {
+	var vInfo volumeInfo
+	vInfo.handle = volumeHandle
+	vInfo.attributes = volumeContext
+	vInfo.syncedAzVolumeAttachments = make(map[string]bool)
+	return &vInfo
+}
 
-			if statusUpdateRequired {
-				azVolumeAttachment.Status.Annotations = azureutils.AddToMap(azVolumeAttachment.Status.Annotations, consts.VolumeAttachmentKey, volumeAttachment.Name)
+func (r *ReconcileAttachDetach) recreateAzVolumeAttachments(ctx context.Context) error {
+	var err error
+	w, _ := workflow.GetWorkflowFromContext(ctx)
 
-				// update status
-				_, err = r.azClient.DiskV1beta2().AzVolumeAttachments(r.config.ObjectNamespace).UpdateStatus(ctx, azVolumeAttachment, metav1.UpdateOptions{})
-				if err != nil {
-					w.Logger().Errorf(err, "failed to update status of AzVolumeAttachment (%s) for volume (%s) and node (%s): %v", azVolumeAttachmentName, *pvName, nodeName, err)
-					return syncedVolumeAttachments, volumesToSync, err
-				}
-			}
+	// try to recreate missing AzVolumeAttachment CRI with Primary role from VolumeAttachment
+	for i := 0; i < maxRetry; i++ {
+		if err = r.recreateAzVolumeAttachmentWithPrimaryRole(ctx); err == nil {
+			break
+		}
+		w.Logger().Errorf(err, "failed to recreate missing AzVolumeAttachment CRI with primary role from VolumeAttachment with retry number (%d).", i)
+	}
 
-			syncedVolumeAttachments[volumeAttachment.Name] = true
+	// try to recreate replica AzVolumeAttachment that had failed to detach for any reason during driver uninstall
+	for i := 0; i < maxRetry; i++ {
+		if err = r.recreateAzVolumeAttachmentWithReplicaRole(ctx); err == nil {
```

Comment:
Maybe it's better to call tryCreateFailedReplicas()?

Reply:
I don't believe so. The `manageReplicas` relies on the current `AzVolumeAttachment` instances to determine how many replicas need to be attached. If we fail to recover an existing replica, `manageReplicas` would not be able to create new replica attachments because the `maxShares` count will be exhausted.

Reply:
`tryCreateFailedReplicas` invokes `manageReplicas` and so also relies on the `AzVolumeAttachments` being in the correct state.